app/models/grda_warehouse/monitoring/tasks/metric_snapshot_collector.rb

```diff
@@ -0,0 +1,380 @@
+###
+# Copyright 2016 - 2025 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+# frozen_string_literal: true
+
+module GrdaWarehouse::Monitoring::Tasks
+  class MetricSnapshotCollector
+    BATCH_SIZE = 5_000
+
+    def self.run_daily_collection(
+      entity_type:,
+      calculation_date: Date.current,
+      entity_ids: nil,
+      metric_names: nil
+    )
+      new(
+        entity_type: entity_type,
+        calculation_date: calculation_date,
+        entity_ids: entity_ids,
+        metric_names: metric_names,
+      ).run
+    end
+
+    def initialize(entity_type:, calculation_date:, entity_ids: nil, metric_names: nil)
+      @entity_type = entity_type
+      @calculation_date = calculation_date
+      @entity_ids = entity_ids
+      @metric_names = metric_names
+      @run_stats = {
+        entities_evaluated: 0,
+        metrics_calculated: 0,
+        snapshots_created: 0,
+        snapshots_updated: 0,
+        errors: 0,
+      }
+    end
+
+    def run
+      # Ensure metric definitions are up-to-date before collection
+      # Uses find_or_create_by! so this is idempotent and preserves user changes
+      GrdaWarehouse::Monitoring::MetricDefinition.maintain!
+
+      run_record = create_run_record
+
+      metrics = load_metrics
+      entities = load_entities
+
+      Rails.logger.info "Collecting #{metrics.count} metrics for #{entities.count} #{@entity_type} records"
+
+      entities.in_groups_of(BATCH_SIZE, false) do |entity_batch|
```

Comment:
`in_groups_of` is an array method right? So this loads the entirety of `entities`. Should we instead use find_in_batches?